### PR TITLE
Initial codebase for Rubrik CDM As Built Report

### PR DESCRIPTION
**Pull request by mwpreston**
Hi all!

Have an initial code base now for the Rubrik CDM As Built Report.  I believe a number of updates will be required to the installation process in the readme once we are published to the PS Gallery, etc, however wanted to get this over to y'all to have a look at before the holidays!  Thanks so much! 🚀

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# Contributing
+Your input is welcome! Contributing to this project is as easy as:
+
+- Reporting a bug
+- Discussing the current state of the code
+- Submitting a fix
+- Proposing new features
+- Creating a new report
+
+When contributing to this repository, please first discuss the change you wish to make via [issue](https://github.com/AsBuiltReport/AsBuiltReport.VMware.vSphere/issues), or [direct message](https://powershell.slack.com/messages/D3MU9DP8S) in the [PowerShell Slack](https://powershell.slack.com) channel before making a change.
```
Comment (tpcarman) [inline CONTRIBUTING.md]: Link to [issue] requires updating to Rubrik repository, currently `(https://github.com/AsBuiltReport/AsBuiltReport.VMware.vSphere/issues)`

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+- It will be a complete piece of work that adds value in some way.
+- It will have a title that reflects the work within, and a summary that helps to understand the context of the change.
+- There will be well written commit messages, with well crafted commits that tell the story of the development of this work.
+- Ideally it will be small and easy to understand. Single commit PRs are usually easy to submit, review, and merge.
+- The code contained within will meet the best practices set by the team wherever possible.
+
+### Submitting pull requests
+1. Fork this repository (AsBuiltReport.VMware.vSphere), or if you are developing a report, fork the specific report repository. The example below uses the main AsBuiltReport.VMware.vSphere repository in the command examples.
+2. Add `https://github.com/AsBuiltReport/AsBuiltReport.VMware.vSphere.git` as a remote named `upstream`.
+    - `git remote add upstream https://github.com/AsBuiltReport/AsBuiltReport.VMware.vSphere.git`
```
Comment (tpcarman) [inline CONTRIBUTING.md]: Information requires updating for Rubrik repository

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    - `git push --force`
+8. Open a Pull Request against the `dev` branch of this repository. We have Pull Requests templates in all repositories for this project. Please follow the template with each Pull Request
+
+Pull requests will be reviewed as soon as possible.
+
+## Any contributions you make will be under the MIT Software License
+In short, when you submit code changes, your submissions are understood to be under the same [MIT License](http://choosealicense.com/licenses/mit/) that covers the project. Feel free to contact the maintainers if that's a concern.
+
+## Report Issues and Bugs
+[GitHub issues](https://github.com/AsBuiltReport/AsBuiltReport.VMware.vSphere/issues) is used to track issues and bugs. Report a bug by opening a new issue, it's that easy!
```
Comment (tpcarman) [inline CONTRIBUTING.md]: Information requires updating for Rubrik repository

File: README.md [hunk 4]
```diff
@@ -1,2 +1,130 @@
...
+There are 4 levels (0/1/3/5) of detail granularity for each section as follows;
+
+| Setting | InfoLevel | Description |
+| ------- | ---- | ----------- |
+| 0 | Disabled | does not collect or display any information
+| 1 | Summary** | provides summarised information for a collection of objects
+| 3 | Detailed | provides detailed information for individual objects
+| 5 | Comprehensive | provides comprehensive information for individual objects, such as advanced configuration settings
+
+***Note*** While you can specify InfoLevels of 2 and 4, they will simply default to the closest defined level below them. IE 2 becomes 1 and 4 becomes 3.
```
Comment (tpcarman) [inline README.md]: Have InfoLevels 2 & 4 been reserved for future use or will there only ever be 4 InfoLevels? InfoLevels do not need to align across reports as each report README and JSON should explain the numbering. Therefore if there will only be 4 InfoLevels, I would suggest making the numbering contiguous, 0-3.

File: AsBuiltReport.Rubrik.CDM.Style.ps1 [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+if ($AsBuiltConfig.Company.FullName) {
+    # Add Company Name if specified
+    Paragraph -Style Title2 $AsBuiltConfig.Company.FullName
+    BlankLine -Count $LineCount
+} else {
+    BlankLine -Count ($LineCount + 1)
+}
+Table -Name 'Cover Page' -List -Style Borderless -Width 0 -Hashtable ([Ordered] @{
+        'Author:' = $AsBuiltConfig.Report.Author
+        'Date:' = Get-Date -Format 'dd MMMM yyyy'
```
Comment (tpcarman) [inline AsBuiltReport.Rubrik.CDM.Style.ps1]: I'm going to use the long date format going forward `(Get-Date).ToLongDateString()`

File: README.md [hunk 6]
```diff
@@ -1,2 +1,130 @@
...
+```powershell
+New-AsBuiltReport -Target 'cluster1.domain.local','cluster2.domain.org' -Username 'administrator@domain.local' -Password 'SuperSecret' -Report Rubrik.CDM -StylePath C:\Scripts\Styles\MyCustomStyle.ps1
+```
+***Note*** the Username/Password combiniation or credentials being passed/used must have administrative rights on both the `cluster1.domain.local` and `cluster2.domain.local` domains.
+
+- Generate HTML & Word reports, attach and send reports via e-mail
+Generate a Rubrik CDM As Built Report for the cluster named 'cluster1.domain.local' using specified credentials. Export report to HTML & DOC formats. Use default report style. Reports are saved to the user profile folder by default. Attach and send reports via e-mail.
+```powershell
+New-AsBuiltReport -Target 'cluster1.domain.local' -Username 'administrator@domain.local' -Password 'domain.local' -Report Vmware.vSphere -Format Html,Word -OutputPath C:\Users\Tim\Documents -SendEmail
```
Comment (tpcarman) [inline README.md]: Report parameter needs updating

**Comment by tquizzle** [comment]
Excited to see this one.

File: CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Rubrik CDM As Built Report Changelog
+
+## 2019-12-09
```
Comment (tpcarman) [inline CHANGELOG.md]: Add version number or `[Unreleased]` - e.g. `[0.1.0] - 2019-12-09` or `[Unreleased] - 2019-12-09`

File: AsBuiltReport.Rubrik.CDM.Style.ps1 [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+Style -Name 'Heading 4' -Size 11 -Color '006A91' 
+Style -Name 'Heading 5' -Size 10 -Color '006A91'
+Style -Name 'Normal' -Size 10 -Color '565656' -Default
+Style -Name 'TOC' -Size 16 -Color '006A91' 
+Style -Name 'TableDefaultHeading' -Size 10 -Color 'FAFAFA' -BackgroundColor '485969'
+Style -Name 'TableDefaultRow' -Size 10 -Color '565656'
+Style -Name 'Critical' -Size 10 -BackgroundColor 'F5DBD9'
+Style -Name 'Warning' -Size 10 -BackgroundColor 'FEF3B5'
+Style -Name 'Info' -Size 10 -BackgroundColor 'E1F1F6'
+Style -Name 'OK' -Size 10 -BackgroundColor 'DFF0D0'
```
Comment (tpcarman) [inline AsBuiltReport.Rubrik.CDM.Style.ps1]: Update colors to reflect Rubrik branding

File: AsBuiltReport.Rubrik.CDM.Style.ps1 [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+# Configure Table Styles
+$TableDefaultProperties = @{
+    Id = 'TableDefault'
+    HeaderStyle = 'TableDefaultHeading'
+    RowStyle = 'TableDefaultRow'
+    BorderColor = '485969'
```
Comment (tpcarman) [inline AsBuiltReport.Rubrik.CDM.Style.ps1]: Update color to reflect Rubrik branding

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+A good quality pull request will have the following characteristics:
+
+- It will be a complete piece of work that adds value in some way.
+- It will have a title that reflects the work within, and a summary that helps to understand the context of the change.
+- There will be well written commit messages, with well crafted commits that tell the story of the development of this work.
+- Ideally it will be small and easy to understand. Single commit PRs are usually easy to submit, review, and merge.
+- The code contained within will meet the best practices set by the team wherever possible.
+
+### Submitting pull requests
+1. Fork this repository (AsBuiltReport.VMware.vSphere), or if you are developing a report, fork the specific report repository. The example below uses the main AsBuiltReport.VMware.vSphere repository in the command examples.
```
Comment (tpcarman) [inline CONTRIBUTING.md]: Update for Rubrik repository

File: Src/Public/Invoke-AsBuiltReport.Rubrik.CDM.ps1 [hunk 11]
```diff
@@ -0,0 +1,1142 @@
+function Invoke-AsBuiltReport.Rubrik.CDM {
+    <#
+    .SYNOPSIS  
+        PowerShell script to document the configuration of Rubrik CDM in Word/HTML/XML/Text formats
+    .DESCRIPTION
+        Documents the configuration of the Rubrik CDM in Word/HTML/XML/Text formats using PScribo.
+    .NOTES
+        Version:        1.0.0
```
Comment (tpcarman) [inline Src/Public/Invoke-AsBuiltReport.Rubrik.CDM.ps1]: I'd suggest version 0.x.x if it's not quite ready for prime time.

**Review by tpcarman — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (mwpreston) [reply CONTRIBUTING.md]: addressed

File: CONTRIBUTING.md (hunk 2 above)
Comment (mwpreston) [reply CONTRIBUTING.md]: addressed

File: CONTRIBUTING.md (hunk 3 above)
Comment (mwpreston) [reply CONTRIBUTING.md]: addressed

File: AsBuiltReport.Rubrik.CDM.Style.ps1 (hunk 5 above)
Comment (mwpreston) [reply AsBuiltReport.Rubrik.CDM.Style.ps1]: addressed

File: README.md (hunk 4 above)
Comment (mwpreston) [reply README.md]: I would like to, if possible to keep 2/4 for future use in the case I ever need to update for particular things.  I have added this information into the documentation so it's a little more clear.

File: CHANGELOG.md (hunk 7 above)
Comment (mwpreston) [reply CHANGELOG.md]: gotchya

File: AsBuiltReport.Rubrik.CDM.Style.ps1 (hunk 8 above)
Comment (mwpreston) [reply AsBuiltReport.Rubrik.CDM.Style.ps1]: done

File: CONTRIBUTING.md (hunk 10 above)
Comment (mwpreston) [reply CONTRIBUTING.md]: done

File: Src/Public/Invoke-AsBuiltReport.Rubrik.CDM.ps1 (hunk 11 above)
Comment (mwpreston) [reply Src/Public/Invoke-AsBuiltReport.Rubrik.CDM.ps1]: done and done - good point!

File: AsBuiltReport.Rubrik.CDM.Style.ps1 (hunk 9 above)
Comment (mwpreston) [reply AsBuiltReport.Rubrik.CDM.Style.ps1]: done

File: README.md (hunk 6 above)
Comment (mwpreston) [reply README.md]: done!

**Comment by mwpreston** [comment]
Hi @tpcarman  - Believe I have addressed all the requests and then some :) 

Thanks

**Review by tpcarman — APPROVED** [review]

**Pull request merged**